src/mlpack/tests/main_tests/linear_svm_test.cpp

```diff
@@ -0,0 +1,893 @@
+  CheckMatricesNotEqual(parameters1, parameters2);
+}
+
+/**
+  * Ensuring that number of classes has some effects on the output.
+ **/
+BOOST_AUTO_TEST_CASE(LinearSVMDiffNumberOfClassesTest)
+{
+  arma::mat trainData;
+  if (!data::Load("iris.csv", trainData))
+    BOOST_FAIL("Cannot load test dataset iris.csv!");
+
+  arma::Row<size_t> trainLabels;
+  if (!data::Load("iris_labels.txt", trainLabels))
+    BOOST_FAIL("Cannot load test dataset iris_labels.txt!");
+
+  SetInputParam("training", trainData);
+  SetInputParam("labels", trainLabels);
+  SetInputParam("number_of_classes", int(0));
+
+  // First solution.
+  mlpackMain();
+
+  // Get the parameters of the output model obtained after first training.
+  const arma::mat parameters1 =
+      std::move(CLI::GetParam<LinearSVM<>*>("output_model")
+                ->Parameters());
+
+  // Reset the settings.
+  bindings::tests::CleanMemory();
+  CLI::ClearSettings();
+  CLI::RestoreSettings(testName);
+
+  SetInputParam("training", std::move(trainData));
+  SetInputParam("labels", std::move(trainLabels));
+  SetInputParam("number_of_classes", int(2));
+
+  // Second solution.
+  mlpackMain();
+
+  // Get the parameters of the output model obtained after second training.
+  const arma::mat &parameters2 =
+      CLI::GetParam<LinearSVM<>*>("output_model")->Parameters();
+
+  // Both solutions should be not equal.
+  CheckMatricesNotEqual(parameters1, parameters2);
+}
+
+/**
+ * The test is only compiled if the user has specified OpenMP to be
+ * used.
+ */
+#ifdef HAS_OPENMP
+/**
+  * Ensuring that step size for optimizer is non negative.
+ **/
+BOOST_AUTO_TEST_CASE(LinearSVMNonNegativeStepSizeTest)
+{
+  arma::mat trainData;
+  if (!data::Load("iris.csv", trainData))
+    BOOST_FAIL("Cannot load test dataset iris.csv!");
+
+  arma::Row<size_t> trainLabels;
+  if (!data::Load("iris_labels.txt", trainLabels))
+    BOOST_FAIL("Cannot load test dataset iris_labels.txt!");
+
+  SetInputParam("training", std::move(trainData));
+  SetInputParam("labels", std::move(trainLabels));
+  SetInputParam("optimizer", std::string("psgd"));
+  SetInputParam("step_size", double(-0.01));
+
+  // Step size for optimizer is negative. It should throw a runtime error.
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+  * Ensuring that shuffle has some effects on the output.
+ **/
+BOOST_AUTO_TEST_CASE(LinearSVMDiffShuffleTest)
+{
+  arma::mat trainData = "2 0 0;"
+                        "0 0 0;"
+                        "0 2 1;"
+                        "1 0 2;"
+                        "0 1 0";
+
+  arma::Row<size_t> trainLabels = "1 0 1";
+
+  SetInputParam("training", trainData);
+  SetInputParam("labels", trainLabels);
+  SetInputParam("optimizer", std::string("psgd"));
+  SetInputParam("number_of_classes", int(2));
+  SetInputParam("shuffle", bool(false));
+
+  // First solution.
+  mlpackMain();
```

Comment:
Note that, for all `psgd` tests, there are two issues that will give different results when they aren't expected:

 1. The random seed.  I think there is no option for `seed` in the program, but we should add one and use `FixedRandomSeed()` for these tests, as parallel SGD may depend on random components (primarily when `shuffle = true`).

 2. The number of cores.  Since parallel SGD allows individual updates in potentially different orders, we should restrict the tests to run on one core only for these tests that are checking if something is different between two runs.

Reply:
I have made the changes as suggested. Is this approach is correct for one core?

Reply:
Yep, this should work fine.  But the last thing to do is ensure that the test only runs on one core. :+1:

Reply:
Probably what's needed is just an ifdef, and if OpenMP is available, then a function to set the number of threads to 1 can be called.